Add support for array_exprt to incremental smt2 decision procedure

thomasspriggs opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Add array_exprt to smt conversion

thomasspriggs committed:
Add regression test

esteffin commented on src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -64,14 +65,42 @@ static std::vector<exprt> gather_dependent_expressions(const exprt &expr)
+  solver_process->send(smt_declare_function_commandt{array_identifier, {}});
+  const std::vector<exprt> &elements = array.operands();
+  const std::size_t index_width =
+    array_sort->index_sort().cast<smt_bit_vector_sortt>()->bit_width();
```
⛏️ 
`array_sort->index_sort()` is never enforced to be of `smt_bit_vector_sortt` type (although we don't use other types).
I suggest adding and `INVARIANT` to ensure the cast is safe

esteffin reviewed (commented):
All good to me.
Just a minor non-blocking comment

thomasspriggs committed:
Avoid assuming that converted arrays have bit vector indices

This means that there will be fewer changes required if we ever change
the encoding of arrays such that they do not have a bit
vector sorted index. It also means that there is no need to add an
invariant for this previously implicitly assumed property.
